src/Microsoft.ComponentDetection.Detectors/nuget/NuGetPackageReferenceFrameworkAwareDetector.cs

```diff
@@ -0,0 +1,231 @@
+
+    private readonly IFileUtilityService fileUtilityService;
+
+    public NuGetPackageReferenceFrameworkAwareDetector(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        IFileUtilityService fileUtilityService,
+        ILogger<NuGetPackageReferenceFrameworkAwareDetector> logger)
+    {
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.Scanner = walkerFactory;
+        this.fileUtilityService = fileUtilityService;
+        this.Logger = logger;
+    }
+
+    public override string Id { get; } = "NuGetPackageReferenceFrameworkAware";
+
+    public override IEnumerable<string> Categories => [Enum.GetName(typeof(DetectorClass), DetectorClass.NuGet)];
+
+    public override IList<string> SearchPatterns { get; } = ["project.assets.json"];
+
+    public override IEnumerable<ComponentType> SupportedComponentTypes { get; } = [ComponentType.NuGet];
+
+    public override int Version { get; } = 1;
+
+    protected override Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs, CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            var lockFile = new LockFileFormat().Read(processRequest.ComponentStream.Stream, processRequest.ComponentStream.Location);
+
+            this.RecordLockfileVersion(lockFile.Version);
+
+            if (lockFile.PackageSpec == null)
+            {
+                throw new FormatException("Lockfile did not contain a PackageSpec");
+            }
+
+            var explicitReferencedDependencies = this.GetTopLevelLibraries(lockFile)
+                .Select(x => this.GetLibraryComponentWithDependencyLookup(lockFile.Libraries, x.Name, x.Version, x.VersionRange))
+                .ToList();
+            var explicitlyReferencedComponentIds =
+                explicitReferencedDependencies
+                    .Select(x => new NuGetComponent(x.Name, x.Version.ToNormalizedString()).Id)
+                    .ToHashSet();
+
+            // Since we report projects as the location, we ignore the passed in single file recorder.
+            var singleFileComponentRecorder = this.ComponentRecorder.CreateSingleFileComponentRecorder(lockFile.PackageSpec.RestoreMetadata.ProjectPath);
+            foreach (var target in lockFile.Targets)
+            {
+                var frameworkPackages = FrameworkPackages.GetFrameworkPackages(target.TargetFramework);
+
+                // This call to GetTargetLibrary is not guarded, because if this can't be resolved then something is fundamentally broken (e.g. an explicit dependency reference not being in the list of libraries)
+                foreach (var library in explicitReferencedDependencies.Select(x => target.GetTargetLibrary(x.Name)).Where(x => x != null))
```

Comment:
There's an existing bug here.  The detector is treating the union of top-level dependencies for all frameworks as top level when evaluating a single target framework.  That's why this `x != null` check is here - it shouldn't be required if the top level dependencies were walked per-framework.

It still ends up walking the dependency graph for that specific target framework so I don't think any dependencies are missed, but it will over-report top-level dependencies.

In the interest of minimizing churn with the existing detector I'll leave this behavior alone.